src/controllers/token.rs

```diff
@@ -114,3 +117,133 @@ pub fn revoke_current(req: &mut dyn RequestExt) -> EndpointResult {
 
     Ok(Response::builder().status(204).body(Body::empty()).unwrap())
 }
+
+/// Verifies that the GitHub signature in request headers is valid
+fn verify_github_signature(req: &dyn RequestExt, json: &[u8]) -> Result<(), Box<dyn AppError>> {
+    // Read and decode request headers
+    let headers = req.headers();
+    let req_key_id = headers
+        .get("GITHUB-PUBLIC-KEY-IDENTIFIER")
+        .ok_or_else(|| bad_request("missing HTTP header: GITHUB-PUBLIC-KEY-IDENTIFIER"))?
+        .to_str()
+        .map_err(|e| bad_request(&format!("failed to decode HTTP header: {e:?}")))?;
+    let sig = headers
+        .get("GITHUB-PUBLIC-KEY-SIGNATURE")
+        .ok_or_else(|| bad_request("missing HTTP header: GITHUB-PUBLIC-KEY-SIGNATURE"))?;
+    let sig = base64::decode(sig)
+        .map_err(|e| bad_request(&format!("failed to decode signature as base64: {e:?}")))?;
+
+    // Fetch list of public keys from GitHub API
+    let app = req.app();
+    let public_keys = app
+        .github
+        .public_keys(&app.config.gh_client_id, &app.config.gh_client_secret)
+        .map_err(|e| bad_request(&format!("failed to fetch GitHub public keys: {e:?}")))?;
+
+    for key in public_keys {
+        if key.key_identifier == req_key_id {
+            if !key.is_current {
+                return Err(bad_request(&format!(
+                    "key id {req_key_id} is not a current key"
+                )));
+            }
+            let key_bytes =
+                key_from_spki(&key).map_err(|_| bad_request("cannot parse public key"))?;
+            let gh_key =
+                signature::UnparsedPublicKey::new(&signature::ECDSA_P256_SHA256_ASN1, &key_bytes);
+
+            return match gh_key.verify(json, &sig) {
+                Ok(v) => {
+                    info!(
+                        "GitHub secret alert request validated with key id {}",
+                        key.key_identifier
+                    );
+                    Ok(v)
+                }
+                Err(e) => Err(bad_request(&format!("invalid signature: {e:?}"))),
+            };
+        }
+    }
+
+    return Err(bad_request(&format!("unknown key id {req_key_id}")));
+}
+
+#[derive(Deserialize, Serialize)]
+struct GitHubSecretAlert {
+    token: String,
+    r#type: String,
+    url: String,
+    source: String,
+}
+
+/// Revokes an API token and notifies the token owner
+fn alert_revoke_token(
+    req: &dyn RequestExt,
+    alert: &GitHubSecretAlert,
+) -> Result<(), Box<dyn AppError>> {
+    let conn = req.db_write()?;
+    // not using ApiToken::find_by_api_token in order to preserve last_used_at
+    let token = api_tokens::table
+        .filter(api_tokens::token.eq(alert.token.as_bytes()))
+        .first::<ApiToken>(&*conn)?;
```

Comment:
what happens if GitHub sends us a token that is a false-positive and not in our database? if I understand correctly it would trigger the "Error revoking API token in GitHub secret alert" error below, but from the docs it sounds like this should not be handled as an error case, but something that can just occasionally happen.

Reply:
You have the option of providing feedback for false positives in the response: https://docs.github.com/en/developers/overview/secret-scanning-partner-program#provide-feedback-for-false-positives

Reply:
@jamisonhyatt FWIW the docs are a bit unclear about whether that payload is meant to be sent as the response. all it says is:

> When you send us feedback

which I initially interpreted as us sending another request to the GitHub API and then wondered what URL to send it to 😅

Reply:
Thanks, it wasn't immediately clear to me either. I've added the feedback in the response now.

Reply:
👋 just noticed this feedback.  I agree, we'll get those updated.